snapcraft/file_utils.py

```diff
@@ -150,18 +150,22 @@ def _search_and_replace_contents(file_path, search_pattern, replacement):
     if os.path.islink(file_path):
         return
 
-    with open(file_path, 'r+') as f:
-        try:
-            original = f.read()
-        except UnicodeDecodeError:
-            # This was probably a binary file. Skip it.
-            return
-
-        replaced = search_pattern.sub(replacement, original)
-        if replaced != original:
-            f.seek(0)
-            f.truncate()
-            f.write(replaced)
+    try:
+        with open(file_path, 'r+') as f:
+            try:
+                original = f.read()
+            except UnicodeDecodeError:
+                # This was probably a binary file. Skip it.
+                return
+
+            replaced = search_pattern.sub(replacement, original)
+            if replaced != original:
+                f.seek(0)
+                f.truncate()
+                f.write(replaced)
+    except PermissionError as e:
+        logger.debug('Unable to open {path} for writing: {error}'.format(
+            path=file_path, error=e))
```

Comment:
The problem with a big try-catch block is that you don't really control which statement is raising the exception. If it's raised in a statement that's not the one you think could raise it, it would be caught anyway and hide a bug.

In this case, I would break python's rule of not asking permission, or open the file twice - one for reading and one for writing in a try-catch. But those are not necessarily cleaner than yours, and as long as the try block doesn't grow with a more complex statement, the risk of a bug is not big. I'm changing my review to leave a +1, so you can keep your code if you like it better than my suggestions.